### ISSUE TITLE
Mopidy segfaults when switching tracks on Arch Linux

### ISSUE THREAD

**Issue by aereaux** [mopidy/mopidy#933]
I tried to use gdb to find out the cause, and it said:

    Program received signal SIGSEGV, Segmentation fault.
    0x00007f5e56cae4d5 in g_mutex_lock () from /usr/lib/libglib-2.0.so.0

**Comment by jodal** [comment]
You're running Arch Linux, right?

Can you provide the output of `mopidy deps`, and also the versions of glib and gstreamer?

**Comment by aereaux** [comment]
Yeah, Arch Linux.  mopidy deps outputs:

    Platform: Linux-3.18.2-1-ck-x86_64-with-glibc2.2.5
    Python: CPython 2.7.9 from /usr/lib/python2.7
    Mopidy: 0.19.5 from /usr/lib/python2.7/site-packages
      setuptools: 11.3.1 from /usr/lib/python2.7/site-packages
      Pykka>=1.1: 1.2.0 from /usr/lib/python2.7/site-packages
      tornado>=2.3: 4.0.2 from /usr/lib/python2.7/site-packages
    GStreamer: 0.10.36.0 from /usr/lib/python2.7/site-packages/gst-0.10/gst
      Detailed information: 
        Python wrapper: gst-python 0.10.22.0
        Relevant elements:
          Found:
            uridecodebin
            souphttpsrc
            appsrc
            alsasink
            osssink
            oss4sink
            pulsesink
            id3demux
            id3v2mux
            lame
            mad
            vorbisdec
            vorbisenc
            vorbisparse
            oggdemux
            oggmux
            oggparse
            flacdec
            flacparse
            shout2send
          Not found:
            flump3dec
            mp3parse

Glib 2.42.1
gstreamer 1.4.5 and
gstreamer0.10 0.10.36, not sure which one you use.

**Comment by rainerborene** [comment]
Same here.

**Comment by reneanda** [comment]
Which client and client version are you using? I have kernel 3.17.6 and no issues with ncmpcpp 0.6.2. Are you using the testing repo?

**Comment by aereaux** [comment]
I am using linux 3.18.2-1-ck, and sonata as the client.

**Comment by reneanda** [comment]
With Sonata 1.6.2.1 I can switch tracks without problems. I am using Spotify, Local and Stream backends.

**Comment by aereaux** [comment]
I have the same version of sonata, and it only segfaults when using the Local backend, not the Spotify backend.

**Comment by rainerborene** [comment]
Linux arch 3.17.6-1-ARCH x86_64 GNU/Linux
Mopidy installed via AUR (version 0.19.5-1)

**Comment by nyc** [comment]
Having the same issue on Arch on kernel 3.18.6, mopidy-0.19.5, and ncmpcpp-0.6.2_pre.

Segfaults with 

```
Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffecdb5700 (LWP 12867)]
0x00007ffff2134015 in g_mutex_lock () from /usr/lib/libglib-2.0.so.0
```

**Comment by kingosticks** [comment]
Someone needs to provide a backtrace

**Comment by nyc** [comment]
I don't know how useful this is, but here's my backtrace. Let me know if I can dig up some more info for you.

```
Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffecdb5700 (LWP 2741)]
0x00007ffff2134015 in g_mutex_lock () from /usr/lib/libglib-2.0.so.0
(gdb) backtrace
#0  0x00007ffff2134015 in g_mutex_lock () from /usr/lib/libglib-2.0.so.0
#1  0x00007fffe6725154 in ?? () from /usr/lib/gstreamer-0.10/libgstcoreelements.so
#2  0x00007fffec37b990 in ?? () from /usr/lib/gstreamer-0.10/libgstplaybin.so
#3  0x00007ffff25c953b in g_cclosure_marshal_VOID__OBJECTv () from /usr/lib/libgobject-2.0.so.0
#4  0x00007ffff25c6484 in ?? () from /usr/lib/libgobject-2.0.so.0
#5  0x00007ffff25e0067 in g_signal_emit_valist () from /usr/lib/libgobject-2.0.so.0
#6  0x00007ffff25e09bf in g_signal_emit () from /usr/lib/libgobject-2.0.so.0
#7  0x00007ffff02484ca in gst_element_remove_pad () from /usr/lib/libgstreamer-0.10.so.0
#8  0x00007ffff25c953b in g_cclosure_marshal_VOID__OBJECTv () from /usr/lib/libgobject-2.0.so.0
#9  0x00007ffff25c6484 in ?? () from /usr/lib/libgobject-2.0.so.0
#10 0x00007ffff25e0067 in g_signal_emit_valist () from /usr/lib/libgobject-2.0.so.0
#11 0x00007ffff25e09bf in g_signal_emit () from /usr/lib/libgobject-2.0.so.0
#12 0x00007ffff02484ca in gst_element_remove_pad () from /usr/lib/libgstreamer-0.10.so.0
#13 0x00007fffe564da44 in ?? () from /usr/lib/gstreamer-0.10/libgstdecodebin2.so
#14 0x00007fffe564db0a in ?? () from /usr/lib/gstreamer-0.10/libgstdecodebin2.so
#15 0x00007fffe564d54a in ?? () from /usr/lib/gstreamer-0.10/libgstdecodebin2.so
#16 0x00007fffe564f165 in ?? () from /usr/lib/gstreamer-0.10/libgstdecodebin2.so
#17 0x00007ffff024ac92 in gst_element_change_state () from /usr/lib/libgstreamer-0.10.so.0
#18 0x00007ffff024b303 in ?? () from /usr/lib/libgstreamer-0.10.so.0
#19 0x00007ffff0239032 in ?? () from /usr/lib/libgstreamer-0.10.so.0
#20 0x00007fffe5659a0d in ?? () from /usr/lib/gstreamer-0.10/libgstdecodebin2.so
#21 0x00007ffff024ac92 in gst_element_change_state () from /usr/lib/libgstreamer-0.10.so.0
#22 0x00007ffff024b303 in ?? () from /usr/lib/libgstreamer-0.10.so.0
#23 0x00007ffff0239032 in ?? () from /usr/lib/libgstreamer-0.10.so.0
#24 0x00007ffff0268a82 in ?? () from /usr/lib/libgstreamer-0.10.so.0
#25 0x00007fffec380053 in ?? () from /usr/lib/gstreamer-0.10/libgstplaybin.so
#26 0x00007ffff024ac92 in gst_element_change_state () from /usr/lib/libgstreamer-0.10.so.0
#27 0x00007ffff024ae82 in gst_element_change_state () from /usr/lib/libgstreamer-0.10.so.0
#28 0x00007ffff024b303 in ?? () from /usr/lib/libgstreamer-0.10.so.0
#29 0x00007ffff0db07f9 in ?? () from /usr/lib/python2.7/site-packages/gst-0.10/gst/_gst.so
#30 0x00007ffff7aef7c4 in PyEval_EvalFrameEx () from /usr/lib/libpython2.7.so.1.0
#31 0x00007ffff7af0af0 in PyEval_EvalCodeEx () from /usr/lib/libpython2.7.so.1.0
#32 0x00007ffff7aeefe4 in PyEval_EvalFrameEx () from /usr/lib/libpython2.7.so.1.0
#33 0x00007ffff7af0af0 in PyEval_EvalCodeEx () from /usr/lib/libpython2.7.so.1.0
#34 0x00007ffff7a7d60d in function_call () from /usr/lib/libpython2.7.so.1.0
#35 0x00007ffff7a588c3 in PyObject_Call () from /usr/lib/libpython2.7.so.1.0
#36 0x00007ffff7aec869 in PyEval_EvalFrameEx () from /usr/lib/libpython2.7.so.1.0
#37 0x00007ffff7aef0ea in PyEval_EvalFrameEx () from /usr/lib/libpython2.7.so.1.0
#38 0x00007ffff7af0af0 in PyEval_EvalCodeEx () from /usr/lib/libpython2.7.so.1.0
#39 0x00007ffff7a7d60d in function_call () from /usr/lib/libpython2.7.so.1.0
#40 0x00007ffff7a588c3 in PyObject_Call () from /usr/lib/libpython2.7.so.1.0
#41 0x00007ffff7aec869 in PyEval_EvalFrameEx () from /usr/lib/libpython2.7.so.1.0
#42 0x00007ffff7aef0ea in PyEval_EvalFrameEx () from /usr/lib/libpython2.7.so.1.0
#43 0x00007ffff7aef0ea in PyEval_EvalFrameEx () from /usr/lib/libpython2.7.so.1.0
#44 0x00007ffff7af0af0 in PyEval_EvalCodeEx () from /usr/lib/libpython2.7.so.1.0
#45 0x00007ffff7a7d52c in function_call () from /usr/lib/libpython2.7.so.1.0
#46 0x00007ffff7a588c3 in PyObject_Call () from /usr/lib/libpython2.7.so.1.0
#47 0x00007ffff7a677ad in instancemethod_call () from /usr/lib/libpython2.7.so.1.0
#48 0x00007ffff7a588c3 in PyObject_Call () from /usr/lib/libpython2.7.so.1.0
#49 0x00007ffff7ae9bb7 in PyEval_CallObjectWithKeywords () from /usr/lib/libpython2.7.so.1.0
#50 0x00007ffff7b1f7c2 in t_bootstrap () from /usr/lib/libpython2.7.so.1.0
#51 0x00007ffff77f8374 in start_thread () from /usr/lib/libpthread.so.0
#52 0x00007ffff753627d in clone () from /usr/lib/libc.so.6
```

**Comment by kingosticks** [comment]
Could you set the environment variable `GST_DEBUG=3` and then reproduce it again (and upload the huge log somewhere)?

**Comment by nyc** [comment]
Not 100% sure if you're talking about some specific log, but i'll link the output of mopidy while running. Interestingly it was a lot harder to make it segfault with the debug option on.

The file is at: goo.gl/iajqfP

**Comment by kingosticks** [comment]
Presumably it was harder since this is some weird race condition between the asynchronous gstreamer state change and something else trying to use a pipeline element that's already been destroyed / hasn't yet been created (because of that state change).

**Comment by kingosticks** [comment]
Does it actually segfault anywhere in that log?

**Comment by nyc** [comment]
It never wrote it to the log, but it segfaulted right at the end.

**Comment by adamcik** [comment]
Is there anything we can actually do about this in Mopidy other than perhaps adding a note in the Arch install docs about glib version issues?

**Comment by jodal** [comment]
Do we know exactly what versions work and which does not? If so, I'll happily add a note to the install docs.

**Comment by Ant59** [comment]
Confirming this is still an issue with both the stable glib and glib-git from the AUR. Mopidy segfaults almost every track change.

**Comment by jodal** [comment]
I know we have several users using Mopidy on Arch and we haven't heard anyone raising this issue for four months now. What's the difference between your systems?

**Comment by Ant59** [comment]
Not sure. Is there any data I can provide to help? It never happened with the Spotify backend, but it happens with Mopidy-YouTube.

**Comment by jodal** [comment]
Spotify is the one doing lots of C stuff, while YouTube is pure Python AFAIK. Thus I think the difference is that Spotify delivers audio data directly to GStreamer, while YouTube is having GStreamering play some URI (like all other extensions, except Spotify).

Can you reproduce with any extension except Spotify?

**Comment by Ant59** [comment]
I've tried with the Podcast and Soundcloud backends and neither seem to have any issues switching tracks. Neither have segfaulted. It seems only to be the Youtube backend.

**Comment by jodal** [comment]
It's possible we're going in the entirely wrong direction here, but I don't have anything else to go on unless someone can compare a working and defunct Arch system...

The difference then is probably the media codecs used, and thus the different libs GStreamer use to decode the audio, depending on the service you play from.

You can try having Mopidy-YouTube request streams of another format to see if the segfault disappears when another library is used for decoding.

1. Find where Mopidy-YouTube is installed: `python -c 'import mopidy_youtube.backend; print mopidy_youtube.backend.__file__'`
2. Edit the file from the previous step.
3. Find the calls to `video.getbest()` and `video.getbestaudio()` on lines 56 and 58 approximately.
4. Change the calls to `video.getbest(preftype='3gp')` and `video.getbestaudio(preftype='3gp')`.
5. Try reproducing the issue.

If you're not able to play with 3gp, you can try "webm" or "flv", ref the usage examples in https://github.com/mps-youtube/pafy/blob/master/README.rst.

**Comment by Ant59** [comment]
I changed it to "webm" and it's now giving me this error instead of the segfault:

```
Aug 20 13:56:23 Eucalyptus mopidy[25709]: g_mutex_clear() called on uninitialised or locked mutex
```

**Comment by joepvd** [comment]
Continuing the discussion from #1284.

For what it is worth, I am having possibly the same issues without using spotify/youtube extensions. A Back Trace:

* Would it be useful if I would try to build gstreamer with debugging symbols? 

* Would it help if I could try and get my gdb working with debugging symbols for python-2.7? 

The backtrace as it currently is:

```
#0  0x00007ffff4e9e3f5 in g_mutex_lock () from /usr/lib/libglib-2.0.so.0
#1  0x00007fffe75d1154 in ?? () from /usr/lib/gstreamer-0.10/libgstcoreelements.so
#2  0x00007fffec6dd990 in ?? () from /usr/lib/gstreamer-0.10/libgstplaybin.so
#3  0x00007ffff5333597 in g_cclosure_marshal_VOID__OBJECTv () from /usr/lib/libgobject-2.0.so.0
#4  0x00007ffff5330524 in ?? () from /usr/lib/libgobject-2.0.so.0
#5  0x00007ffff5349f97 in g_signal_emit_valist () from /usr/lib/libgobject-2.0.so.0
#6  0x00007ffff534a8ef in g_signal_emit () from /usr/lib/libgobject-2.0.so.0
#7  0x00007fffefd2d4ca in gst_element_remove_pad () from /usr/lib/libgstreamer-0.10.so.0
#8  0x00007ffff5333597 in g_cclosure_marshal_VOID__OBJECTv () from /usr/lib/libgobject-2.0.so.0
#9  0x00007ffff5330524 in ?? () from /usr/lib/libgobject-2.0.so.0
#10 0x00007ffff5349f97 in g_signal_emit_valist () from /usr/lib/libgobject-2.0.so.0
#11 0x00007ffff534a8ef in g_signal_emit () from /usr/lib/libgobject-2.0.so.0
#12 0x00007fffefd2d4ca in gst_element_remove_pad () from /usr/lib/libgstreamer-0.10.so.0
#13 0x00007fffd95e5a44 in ?? () from /usr/lib/gstreamer-0.10/libgstdecodebin2.so
#14 0x00007fffd95e7165 in ?? () from /usr/lib/gstreamer-0.10/libgstdecodebin2.so
#15 0x00007fffefd2fc92 in gst_element_change_state () from /usr/lib/libgstreamer-0.10.so.0
#16 0x00007fffefd30303 in ?? () from /usr/lib/libgstreamer-0.10.so.0
#17 0x00007fffefd1e032 in ?? () from /usr/lib/libgstreamer-0.10.so.0
#18 0x00007fffd95f1a0d in ?? () from /usr/lib/gstreamer-0.10/libgstdecodebin2.so
#19 0x00007fffefd2fc92 in gst_element_change_state () from /usr/lib/libgstreamer-0.10.so.0
#20 0x00007fffefd30303 in ?? () from /usr/lib/libgstreamer-0.10.so.0
#21 0x00007fffefd1e032 in ?? () from /usr/lib/libgstreamer-0.10.so.0
#22 0x00007fffefd4da82 in ?? () from /usr/lib/libgstreamer-0.10.so.0
#23 0x00007fffec6e2053 in ?? () from /usr/lib/gstreamer-0.10/libgstplaybin.so
#24 0x00007fffefd2fc92 in gst_element_change_state () from /usr/lib/libgstreamer-0.10.so.0
#25 0x00007fffefd2fe82 in gst_element_change_state () from /usr/lib/libgstreamer-0.10.so.0
#26 0x00007fffefd30303 in ?? () from /usr/lib/libgstreamer-0.10.so.0
#27 0x00007ffff08957f9 in ?? () from /usr/lib/python2.7/site-packages/gst-0.10/gst/_gst.so
#28 0x00007ffff7af4b3f in PyEval_EvalFrameEx () from /usr/lib/libpython2.7.so.1.0
#29 0x00007ffff7af5abc in PyEval_EvalCodeEx () from /usr/lib/libpython2.7.so.1.0
#30 0x00007ffff7af4786 in PyEval_EvalFrameEx () from /usr/lib/libpython2.7.so.1.0
#31 0x00007ffff7af5abc in PyEval_EvalCodeEx () from /usr/lib/libpython2.7.so.1.0
#32 0x00007ffff7a824ad in function_call () from /usr/lib/libpython2.7.so.1.0
#33 0x00007ffff7a5d9e3 in PyObject_Call () from /usr/lib/libpython2.7.so.1.0
#34 0x00007ffff7af2671 in PyEval_EvalFrameEx () from /usr/lib/libpython2.7.so.1.0
#35 0x00007ffff7af4833 in PyEval_EvalFrameEx () from /usr/lib/libpython2.7.so.1.0
#36 0x00007ffff7af5abc in PyEval_EvalCodeEx () from /usr/lib/libpython2.7.so.1.0
#37 0x00007ffff7a824ad in function_call () from /usr/lib/libpython2.7.so.1.0
#38 0x00007ffff7a5d9e3 in PyObject_Call () from /usr/lib/libpython2.7.so.1.0
#39 0x00007ffff7af2671 in PyEval_EvalFrameEx () from /usr/lib/libpython2.7.so.1.0
#40 0x00007ffff7af4833 in PyEval_EvalFrameEx () from /usr/lib/libpython2.7.so.1.0
#41 0x00007ffff7af4833 in PyEval_EvalFrameEx () from /usr/lib/libpython2.7.so.1.0
#42 0x00007ffff7af5abc in PyEval_EvalCodeEx () from /usr/lib/libpython2.7.so.1.0
#43 0x00007ffff7a823cc in function_call () from /usr/lib/libpython2.7.so.1.0
#44 0x00007ffff7a5d9e3 in PyObject_Call () from /usr/lib/libpython2.7.so.1.0
#45 0x00007ffff7a6c7c7 in instancemethod_call () from /usr/lib/libpython2.7.so.1.0
#46 0x00007ffff7a5d9e3 in PyObject_Call () from /usr/lib/libpython2.7.so.1.0
#47 0x00007ffff7aeeed7 in PyEval_CallObjectWithKeywords () from /usr/lib/libpython2.7.so.1.0
#48 0x00007ffff7b25142 in t_bootstrap () from /usr/lib/libpython2.7.so.1.0
#49 0x00007ffff77fd4a4 in start_thread () from /usr/lib/libpthread.so.0
#50 0x00007ffff753b13d in clone () from /usr/lib/libc.so.6
```

**Comment by adamcik** [comment]
To be honest we probably won't be able to get his fixed given GStreamer 0.10. It's no longer maintained, so the fix is basically to finish our planed migration to 1.0

**Comment by mguentner** [comment]
This has also been reported on discuss: https://discuss.mopidy.com/t/mopidy-crashes-on-use-of-next-song/860

I have created a PR with a modified PKGBUILD + a patch that disables the priority inheritance support for pulseaudio:
https://github.com/archlinuxarm/PKGBUILDs/pull/1268

After installing the new `.pkg`, switching tracks does not result in a segfault.

**Comment by adamcik** [comment]
So this now has a known fix outside of mopidy, any reason to keep this bug open?

**Comment by jodal** [comment]
No, I don't think so.

**Comment by NicoHood** [comment]
I am coming from this forum thread:
https://discuss.mopidy.com/t/mopidy-crashes-on-use-of-next-song/860

I do not know why it linked to this issue, it looks totally different, but I did not yet want to open a new issue.

I installed mopidy from the arch repo for raspi2. I got the same error as described in the post, but no segfault (or at least its not named as segfault). The last comment says it got fixed with pip. So I assume the current git master has a fix and its only waiting for a new release tag?
